### PR TITLE
Support using BCELoss for training regression models in AutoMM + Support MinMaxScaler

### PR DESCRIPTION
**Pull request by Linuxdex**
*Issue #, if available:*
Adding BCEloss into automm regression. It is a special method for dealing with regression problems.
*Description of changes:*
Add BCEloss. It will be active if the problem_type = binary and the eval_metrics belong to regression.
This requires the label to be converted to [0,1].
So the default standardization should be modified. Now users can choose "minmaxscaler", "standardscaler" or nothing to do in the data default.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: text/src/autogluon/text/automm/configs/data/default.yaml [hunk 1]
```diff
@@ -10,6 +10,8 @@ data:
     convert_to_text: False  # Whether to convert the feature to text.
     scaler_with_mean: True  # Whether to normalize with mean.
     scaler_with_std: True  # Whether to normalize with std.
+  label:
+    numerical_label_preprocessing:  #The mode of label preprocessing for . Support "standardscaler" or "minmaxscaler" or null.
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/configs/data/default.yaml:14]: Change `#The mode` to `# The mode`.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 def get_loss_func(
         problem_type: str,
         mixup_active: bool,
+        bce_type: bool,
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/optimization/utils.py:26]: How about to change `bce_type` to `loss_function` and use None by default. The loss function can then be `bce` and will be used if it's regression problem.

Also, we need to revise the docstring below (add mixup_active and loss_function to parameters).

File: text/src/autogluon/text/automm/predictor.py [hunk 3]
```diff
@@ -414,16 +418,17 @@ def fit(
             pos_label=pos_label,
         )
 
-        mixup_active, mixup_fn = get_mixup(model_config=OmegaConf.select(config,'model'),
-                                     mixup_config=OmegaConf.select(config,'data.mixup'),
-                                     num_classes=output_shape,
+        mixup_active, mixup_fn = get_mixup(
+            model_config=OmegaConf.select(config,'model'),
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/predictor.py:422]: Add space `(config,'model')` should be `(config, 'model')`.

File: text/src/autogluon/text/automm/predictor.py [hunk 4]
```diff
@@ -147,6 +147,8 @@ def __init__(
             eval_metric = eval_metric.name
 
         if eval_metric is not None and eval_metric.lower() in ["rmse", "r2", "pearsonr", "spearmanr"]:
+            if problem_type == BINARY:
+                bce_regression_state = True
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/predictor.py:151]: Cannot understand the logic.

File: text/src/autogluon/text/automm/predictor.py [hunk 5]
```diff
@@ -1032,6 +1037,13 @@ def _logits_to_prob(logits: torch.Tensor):
         prob = prob.detach().cpu().float().numpy()
         return prob
 
+    @staticmethod
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/predictor.py:1040]: Instead of using a static method, let's make it a function and put it in utils or other places.

**Comment by sxjscience** [comment]
I think we may later turn on `MinMaxScale` if it's regression problem and the user is using bce loss.

File: text/src/autogluon/text/automm/predictor.py (hunk 4 above)
Comment (Linuxdex) [reply text/src/autogluon/text/automm/predictor.py:151]: In my original design, this mode is determine to introduce binary into regression. So users treat it as a binary problem (but regression indeed) and the eval metrics may be regression.

**Comment by szha** [comment]
Job PR-1777-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1777/3/index.html

**Comment by szha** [comment]
Job PR-1777-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1777/5/index.html

**Comment by szha** [comment]
Job PR-1777-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1777/6/index.html

File: text/src/autogluon/text/automm/configs/optimization/adamw.yaml [hunk 6]
```diff
@@ -16,3 +16,6 @@ optimization:
   top_k_average_method: "greedy_soup"  # We support averaging method described in https://arxiv.org/pdf/2203.05482.pdf.
                                        # Currently support "uniform_soup", "greedy_soup", and "best".
   efficient_finetune: null  # Can be 'bit_fit' (only finetune bias), 'norm_fit' (finetune the normalization terms + bias terms), or null
+  loss_function: "BCEWithLogitsLoss" # The replaced loss for regresssion. Can only support loss function in torch.nn.
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/configs/optimization/adamw.yaml:19]: Use an empty value by default and only specify the value in test-cases.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 7]
```diff
@@ -477,3 +483,34 @@ def apply_layerwise_lr_decay(
         parameter_group_names[group_name]["params"].append(name)
 
     return list(parameter_group_vars.values())
+
+
+def modify_config_with_loss_func(
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/optimization/utils.py:488]: Consider to use a name like 
```
def config_update_loss_function(...):
```

File: text/src/autogluon/text/automm/data/preprocess_dataframe.py [hunk 8]
```diff
@@ -435,6 +448,9 @@ def transform_prediction(
         -------
         Predicted labels ready to compute metric scores.
         """
+        if OmegaConf.select(self.config, "label.sigmoid"):
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/data/preprocess_dataframe.py:451]: There is no "label.sigmoid" in the config.

File: text/src/autogluon/text/automm/configs/optimization/adamw.yaml (hunk 6 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/configs/optimization/adamw.yaml:19]: Actually, we can also pick "auto" by default, which is an indicator that it should be updated by the system.

**Comment by sxjscience** [comment]
We recently enabled black to check the coding style so you will have to rebase the code with the latest master.

**Comment by szha** [comment]
Job PR-1777-22 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1777/22/index.html

**Review by sxjscience — APPROVED** [review]
LGTM

File: text/tests/unittests/automm/test_metrics.py [hunk 9]
```diff
@@ -40,3 +40,35 @@ def test_cross_entropy(metric_name, class_num):
...
+        problem_type=problem_type,
+        mixup_active=False,
+        loss_func_replaced=loss_func_replaced,
+    )
+
+    score1 = loss_func(input=preds, target=targets)
+    preds = preds.sigmoid()
+    bceloss = torch.nn.BCELoss()
+    score2 = bceloss(input=preds, target=targets)
+    assert pytest.approx(score1, 1e-6) == score2
```
Comment (bryanyzhu) [inline text/tests/unittests/automm/test_metrics.py:74]: Maybe need a newline for the end of the file.

**Review by bryanyzhu — APPROVED** [review]

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 10]
```diff
@@ -57,7 +59,13 @@ def get_loss_func(
...
             loss_func = nn.CrossEntropyLoss()
     elif problem_type == REGRESSION:
-        loss_func = nn.MSELoss()
+        if loss_func_replaced is not None:
+            if "bcewithlogitsloss" in loss_func_replaced.lower():
+                loss_func = nn.BCEWithLogitsLoss()
+            else:
+                loss_func = nn.MSELoss()
+        else:
+            loss_func = nn.MSELoss()
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:68]: This implementation restricted using bce to only regression tasks. We can improve this code. Keep the original code and Insert the:
```
if loss_func_replaced is not None and "bcewithlogitsloss" in loss_func_replaced.lower():
    loss_func = nn.BCEWithLogitsLoss()
```
at the end of this function. In this way, users can also use bce loss for classification tasks.

Comment (sxjscience) [reply text/src/autogluon/text/automm/optimization/utils.py:68]: Yeah, how about to revise it in a later PR. I think BCE for classification has not been tested. We can fix the logic later.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 11]
```diff
@@ -38,6 +39,7 @@
 def get_loss_func(
     problem_type: str,
     mixup_active: bool,
+    loss_func_replaced: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/utils.py:42]: How about changing `loss_func_replaced`  to `loss_func_name`?

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 12]
```diff
@@ -480,3 +488,30 @@ def apply_layerwise_lr_decay(
         parameter_group_names[group_name]["params"].append(name)
 
     return list(parameter_group_vars.values())
+
+
+def config_update_loss_func(
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:497]: Could we change `config_update_loss_func` to `update_config_by_rules`? I guess we will have other similar needs to update config based on our pre-defined rules. So, we can make this function name more general and add other similar stuff into this function in future.

File: text/src/autogluon/text/automm/predictor.py [hunk 13]
```diff
@@ -452,7 +458,8 @@ def fit(
         self._df_preprocessor = df_preprocessor
         self._data_processors = data_processors
         self._model = model
-        self.mixup_fn = mixup_fn
+        self._mixup_fn = mixup_fn
+        self._loss_func = loss_func
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:462]: There is no need to have these two attributes since they are not saved.

Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:462]: Yeah, I just see `self._loss_func` is used below, but `self._mixup_fn` seem not necessary.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 14]
```diff
@@ -480,3 +492,30 @@ def apply_layerwise_lr_decay(
...
+    config
+        The config of the project. It is a Dictconfig object.
+
+    Returns
+    -------
+    The modified config.
+    """
+    if problem_type == REGRESSION and OmegaConf.select(config, "optimization.loss_function") is not None:
+        loss_func = config.optimization.loss_function
+        if "bceloss" in loss_func.lower() or "bcewithlogitsloss" in loss_func.lower():
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:518]: Is `bceloss` a valid choice? I see only `bcewithlogitsloss` are allowed in `get_loss_func`.

File: text/src/autogluon/text/automm/optimization/utils.py (hunk 10 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/utils.py]: The current implementations in `transform_prediction` and `_compute_metric_score` seem not to support BCE for classification. Considering this limitation, we should raise error if detecting BCE in classification tasks.

**Review by zhiqiangdon — APPROVED** [review]
Thanks for adding BCE support for regression!
Since the current BCE is crafted for regression, we need to raise warnings/errors if users specify BCE for classification tasks.

**Comment by szha** [comment]
Job PR-1777-36 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1777/36/index.html
